### PR TITLE
 ARROW-17980: [C++] As-of-Join Substrait extension #14385 

### PR DESCRIPTION
**Pull request by westonpace**

### CODE REVIEW

File: cpp/cmake_modules/ThirdpartyToolchain.cmake [hunk 1]
```diff
@@ -1767,6 +1771,29 @@ macro(build_substrait)
 
     list(APPEND SUBSTRAIT_SOURCES "${SUBSTRAIT_PROTO_GEN}.cc")
   endforeach()
+  message("SOURCE DIR2 IS ${SOURCE_DIR} AND ${CMAKE_SOURCE_DIR} AND ${ARROW_SUBSTRAIT_PROTOS_DIR}"
+  )
```
Comment (westonpace) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: Let's remove this cleanup if its ready

File: cpp/src/arrow/compute/exec/options.h [hunk 2]
```diff
@@ -393,22 +393,35 @@ class ARROW_EXPORT HashJoinNodeOptions : public ExecNodeOptions {
...
-  /// All inputs tables must be sorted by the "on" key. Must be a single field of a common
-  /// type. Inexact match is used on the "on" key. i.e., a row is considered match iff
-  /// left_on - tolerance <= right_on <= left_on.
-  /// Currently, the "on" key must be of an integer, date, or timestamp type.
-  FieldRef on_key;
-  /// \brief "by" key for the join.
+  /// The keys must be consistent across the input tables:
+  /// Each "on" key must refer to a field of the same type and units across the tables.
+  /// Each "by" key must refer to a list of fields of the same types across the tables.
+  struct Keys {
```
Comment (westonpace) [inline cpp/src/arrow/compute/exec/options.h]: So there is one `Keys` element per input?  Do they all need to have the exact same number of `by_key` keys?

File: cpp/src/arrow/engine/substrait/options.cc [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      arrow::substrait::AsOfJoinRel as_of_join_rel;
+      rel.UnpackTo(&as_of_join_rel);
+      return MakeAsOfJoinRel(inputs, as_of_join_rel, ext_set);
+    }
+    return Status::NotImplemented("Unrecognized extension in Susbstrait plan: ",
+                                  rel.DebugString());
+  }
+
+ private:
+  Result<DeclarationInfo> MakeAsOfJoinRel(
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/options.cc:48]: It's a bit weird to have this in `options.cc`.  I think, when we looking at merging into arrow, we will want to find a better home for this.  Perhaps a `custom_relations.cc` or something like that.

File: cpp/src/arrow/engine/substrait/serde_test.cc [hunk 4]
```diff
@@ -3279,6 +3299,195 @@ TEST(Substrait, IsthmusPlan) {
                        *compute::default_exec_context(), buf, {}, conversion_options);
 }
 
+TEST(Substrait, PlanWithExtension) {
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde_test.cc]: Can we change the name of this test to reflect that it has something to do with asof join?

**Review by westonpace — COMMENTED** [review]
This looks pretty straightforward.  I'll review the substrait changes more closely when we can figure out how to get rid of the `::substrait` change (I think Anja was making progress on this)?

It looks like the major changes are:

 * Change as-of join node so that it can have unique key indices per table
 * Pull the MakeOutputSchema method out into its own header file so it can be referenced by the substrait mapping
 * Add substrait mapping for asof join node

Is that correct?

@rtpsw

**Comment by rtpsw** [comment]
@westonpace, I see this is a PR on your personal GitHub space. Is this intended?

**Comment by westonpace** [comment]
> @westonpace, I see this is a PR on your personal GitHub space. Is this intended?

I think so.  My understanding is that we are going to do some rapid prototyping on this branch.

**Comment by rtpsw** [comment]
> I'll review the substrait changes more closely when we can figure out how to get rid of the `::substrait` change (I think Anja was making progress on this)?

I discussed this issue with @icexelloss. Instead of this `::substrait` change, a more localized change may be possible by a `using` statement for the namespace. I'll look into this.

File: cpp/src/arrow/compute/exec/options.h (hunk 2 above)
Comment (rtpsw) [reply cpp/src/arrow/compute/exec/options.h]: Yes and yes.

File: cpp/cmake_modules/ThirdpartyToolchain.cmake (hunk 1 above)
Comment (rtpsw) [reply cpp/cmake_modules/ThirdpartyToolchain.cmake]: Will do.

File: cpp/src/arrow/engine/substrait/options.cc (hunk 3 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/options.cc:48]: Will do.

File: cpp/src/arrow/engine/substrait/serde_test.cc (hunk 4 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/serde_test.cc]: Will do.

**Pull request closed without merging**